config/modules/dns/master/default.nix

```diff
@@ -0,0 +1,132 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.hackint.dns.master;
+
+  zone = pkgs.writeText "${cfg.zones.dynamic.domain}.zone" ''
+    $ORIGIN ${cfg.zones.dynamic.domain}.
+    $TTL 60
+
+    @ SOA ${head cfg.nameservers} mail.hackint.org 0 200 300 1209600 300
+
+    ${concatMapStringsSep "\n"
+      (server: "@ NS ${server}")
+      cfg.nameservers}
+  '';
+
+in
+{
+  options.hackint.dns.master = with types; {
+    enable = mkEnableOption "the DNS master";
+
+    upstreams = mkOption {
+      type = listOf str;
+      description = ''
+        Remote adrresses of upstream servers.
+      '';
+    };
+
+    nameservers = mkOption {
+      type = listOf str;
+      description = ''
+        FQDNs of public serving servers.
+      '';
+    };
+
+    zones.dynamic.domain = mkOption {
+      type = str;
+      description = ''
+        The zone name.
+      '';
+      default = "dyn.${config.networking.domain}";
+    };
+
+    zones.dynamic.networks = mkOption {
+      type = listOf str;
+      description = ''
+        Hosts and networks allowed to do dynamic updates.
+      '';
+    };
+
+    zones.dynamic.secret = mkOption {
+      type = str;
+      description = ''
+        Secret TSIG key for dynamic updates.
+      '';
+    };
+
+    zones.static = mkOption {
+      type = attrsOf path;
+      description = ''
+        Static zones to server.
+      '';
+      default = { };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    # Let systemd-resolved not listen on 127.0.0.53:53 to avoid conflicts with
+    # kresd listening on wildcard.
+    services.resolved.extraConfig = ''
+      DNSStubListener=no
+    '';
+
+    services.knot = {
+      enable = true;
+
+      extraConfig = ''
+        server:
+          listen: [ "0.0.0.0@53", "::@53" ]
+
+        key:
+          - id: "update"
+            algorithm: hmac-sha512
+            secret: "${cfg.zones.dynamic.secret}"
+
+        remote:
+          - id: "upstream"
+            address: [${concatMapStringsSep "," (s: "\"${s}@53\"") cfg.upstreams}]
+
+        acl:
+          - id: "upstream"
+            address: [${concatMapStringsSep "," (s: "\"${s}\"") cfg.upstreams}]
+            action: transfer
+
+          - id: "update"
+            address: [${concatMapStringsSep "," (s: "\"${s}\"") cfg.zones.dynamic.networks}]
+            action: update
+            update-type: TXT
+            key: [ "update" ]
+
+        policy:
+          - id: "default"
+            algorithm: ed25519
+            cds-cdnskey-publish: always
```

Comment:
That would be awesome if our public nameservers supported that.

Reply:
Not only our public nameservers but also the according registries for the TLDs. But it does not hurt to have it there.